lib/tapioca/compilers/dsl/smart_properties.rb

```diff
@@ -154,10 +158,10 @@ def type_for(property)
             "T.untyped"
           end
 
-          required_attr = property.instance_variable_get(:@required)
-          required = !required_attr.is_a?(Proc) && !!required_attr
-          property_required = type == "T.untyped" || required
-          type = "T.nilable(#{type})" unless property_required
+          not_required = Proc === required || !required
+          nilable_default = Proc === default || default.nil?
```

Comment:
That would still make `default.nil?` be `true` since `T.let` just returns the argument value. If there are proxy objects that wrap a nil value, then this would be problematic, though. But, I think that is a super rare thing, and this heuristic should be fine for most purposes.